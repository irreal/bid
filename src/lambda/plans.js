const mongoose = require("mongoose");
const planSchema = new mongoose.Schema({
  title: String,
  period_start: Date,
  period_end: Date,
  date_added: Date,
  percent_complete: Number,
  created_by: String,
  allowed_roles: [String],
  allowed_users: [String]
});
var Plan = mongoose.model("plans", planSchema);
export async function handler(event, context, callback) {
  const { user } = context.clientContext;
  if (!user) {
    callback(null, {
      statusCode: 401,
      body: "missing client context"
    });
    return;
  }
  try {
    const uri = process.env.MONGO_CONNECTION;
    await mongoose.connect(uri, {
      useNewUrlParser: true
    });

    if (event.httpMethod === "POST") {
      try {
        const params = JSON.parse(event.body);
        console.log("params: ", params);
        if (!params.title) {
          throw new Error("missing parameter title");
        }
        if (!params.period_start) {
          throw new Error("missing parameter period start");
        }
        if (!params.period_end) {
          throw new Error("missing parameter period end");
        }

        var newPlan = new Plan({
          title: params.title,
          percent_complete: 0,
          period_start: params.period_start,
          period_end: params.period_end,
          date_added: new Date(),
          created_by: user.email
        });
        await newPlan.save();
        callback(null, {
          statusCode: 201,
          body: '"ok"'
        });
      } catch (err) {
        console.log("evo gre≈°ke, ", err);
        callback(null, {
          statusCode: 400,
          body: `"Invalid request\n${err.toString()}"`
        });
      }
      return;
    }
    if (event.httpMethod !== "GET") {
      callback(null, {
        statusCode: 405
      });
      return;
    }

    let filter = {};
    if (!user.app_metadata.roles.includes("admin")) {
      filter = {
        $or: [
          {
            created_by: user.email
          },
          {
            allowed_roles: {
              $in: user.app_metadata.roles
            }
          },
          {
            allowed_users: user.email
          }
        ]
      };
    }
    const results = await Plan.find(
      filter,
      "_id title percent_complete period_start period_end created_by"
    );
    await mongoose.connection.close();
    callback(null, {
      statusCode: 200,
      body: JSON.stringify(results)
    });
  } catch (err) {
    await mongoose.connection.close();
    console.log("error accessing plans ", err);
    callback(err);
  }
}
