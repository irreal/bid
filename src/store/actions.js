export default {
  updateUser: ({ commit }, payload) => {
    commit("setUser", payload.currentUser);
  }
};
