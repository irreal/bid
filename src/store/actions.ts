import { ActionTree } from 'vuex';
import State from './state';

const actions: ActionTree<State, State> = {
    //   updateColor({ commit }: ActionContext<State, State>, newColor: string): void {
    //     commit(COLOR, newColor);
    //   },
    };

export default actions;
