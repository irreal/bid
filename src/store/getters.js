export default {
  getUserStatus: state => !!state.user,
  getUser: state => state.user
};
