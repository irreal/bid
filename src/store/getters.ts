import { GetterTree } from 'vuex';
import State from './state';
import Plan from '@/models/plan';

const getters: GetterTree<State, State> = {
    plans(state: State): Plan[] {
      return state.plans;
    },
    // colour(state: State): string {
    // 	return state.color;
    // },
  };

export default getters;
