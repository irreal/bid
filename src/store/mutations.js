export default {
  setUser: (state, currentUser) => {
    state.user = currentUser;
  }
};
