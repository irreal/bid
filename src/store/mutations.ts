import { MutationTree } from 'vuex';
import State from './state';

const mutations: MutationTree<State> = {
    removePlans(state: State) {
      state.plans = [];
    },
    // color(state: State, newColor: string): void {
    // 	state.color = newColor;
    // }
  };

export default mutations;
