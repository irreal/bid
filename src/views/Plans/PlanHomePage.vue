<template>
  <div>
    <div class="plan-container d-flex flex-column">
      <apexchart
        type="bar"
        height="100%"
        :options="chartOptions"
        :series="chartData"
      />
    </div>
  </div>
</template>
<style scoped lang="scss">
h1 {
  margin-left: 25px;
}
.plan-container {
  height: 85vh;
}
</style>

<script>
import chartOptions from "./PlanChartOptions";

export default {
  name: "PlanHomePage",
  computed: {
    chartData() {
      return [
        {
          name: "Realizacija",
          data:
            this.plans && this.plans.length
              ? this.plans.map(p => p.percent_complete)
              : []
        }
      ];
    },
    chartOptions() {
      const options = { ...chartOptions };
      options.chart = {
        ...chartOptions.chart,
        events: { dataPointSelection: this.clickedItem }
      };
      options.xaxis.categories =
        this.plans && this.plans.length ? this.plans.map(p => p.title) : [];
      return options;
    },
    plans() {
      return this.$store.getters["plans/all"];
    }
  },
  data() {
    return {};
  },
  mounted() {
    this.$store.dispatch("plans/load");
  },
  methods: {
    clickedItem(_, __, { selectedDataPoints }) {
      if (selectedDataPoints[0][0] !== undefined) {
        const plan = this.plans[selectedDataPoints[0][0]];
        if (plan) {
          this.$router.push({
            name: "plan-detail",
            params: { planId: plan._id }
          });
        }
      }
    }
  }
};
</script>
