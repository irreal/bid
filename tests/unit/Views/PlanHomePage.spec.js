import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import PlanHomePage from "@/views/Plans/PlanHomePage.vue";
import Vuetify from "vuetify";
import Vue from "vue";
import initialState from "@/store/state";
import actions from "@/store/actions";
import getters from "@/store/getters";
import mutations from "@/store/mutations";
import plans from "@/store/plans/plans-store";
import userFixture from "../fixtures/user";
import plansFixture from "../fixtures/plans";
import VueApexCharts from "vue-apexcharts";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueApexCharts);
localVue.component("apexchart", VueApexCharts);
Vue.use(Vuetify);

describe("Plan Home Page", () => {
  let state;

  const defaultStoreObject = () => ({
    state,
    actions,
    getters,
    mutations,
    modules: {
      plans
    }
  });
  const build = (storeObject = defaultStoreObject()) => {
    const store = new Vuex.Store(storeObject);

    const plan = shallowMount(PlanHomePage, {
      localVue,
      store
    });

    return {
      plan
    };
  };

  beforeEach(() => {
    state = { ...initialState };
  });

  it("renders the component", () => {
    // arrange
    const { plan } = build();
    // assert
    expect(plan.html()).toMatchSnapshot();
  });

  it("renders the component with a logged in user", () => {
    // arrange
    state.user = userFixture;
    const { plan } = build();
    // assert
    expect(plan.html()).toMatchSnapshot();
  });

  it("exposes empty chart data if there are no plans", () => {
    const { plan } = build();
    expect(plan.vm.chartData[0].data).toEqual([]);

    expect(plan.vm.chartOptions.xaxis.categories).toEqual([]);
  });

  it("exposes plans through chart properties", () => {
    const { plan } = build();
    plan.setData({
      plans: plansFixture
    });
    expect(plan.vm.chartData[0].data).toEqual(
      plansFixture.map(pf => pf.percent_complete)
    );

    expect(plan.vm.chartOptions.xaxis.categories).toEqual(
      plansFixture.map(pf => pf.title)
    );
  });

  it("reacts to selecting a plan by navigating", () => {
    const { plan } = build();
    plan.vm.$router = { push: jest.fn() };
    plan.setData({
      plans: plansFixture
    });
    plan.vm.clickedItem(null, null, { selectedDataPoints: [[0]] });
    expect(plan.vm.$router.push).toBeCalledWith({
      name: "plan-detail",
      params: { planId: plansFixture[0]._id }
    });
  });
  it("reacts to selecting a plan by doing nothing if plan not found", () => {
    const { plan } = build();
    plan.vm.$router = { push: jest.fn() };
    plan.vm.clickedItem(null, null, { selectedDataPoints: [[0]] });
    expect(plan.vm.$router.push).not.toBeCalled();
  });
  it("reacts to selecting a plan by doing nothing if selected data point is not provided ", () => {
    const { plan } = build();
    plan.vm.$router = { push: jest.fn() };
    plan.vm.clickedItem(null, null, { selectedDataPoints: [[undefined]] });
    expect(plan.vm.$router.push).not.toBeCalled();
  });
});
