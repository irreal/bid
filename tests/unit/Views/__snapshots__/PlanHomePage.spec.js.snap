// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan Home Page renders the component 1`] = `
<div class="fill-height plan-page">
  <div class="plan-container d-flex flex-column">
    <apexchart-stub options="[object Object]" type="bar" series="[object Object]" width="100%" height="100%" data-test="plans-chart"></apexchart-stub>
  </div>
</div>
`;

exports[`Plan Home Page renders the component with a logged in user 1`] = `
<div class="fill-height plan-page">
  <div class="plan-container d-flex flex-column">
    <apexchart-stub options="[object Object]" type="bar" series="[object Object]" width="100%" height="100%" data-test="plans-chart"></apexchart-stub>
  </div>
</div>
`;
